core/common/src/main/java/alluxio/conf/PathConfiguration.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.conf;
+
+import alluxio.AlluxioURI;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Class used to resolve path level configuration.
+ *
+ * Preference for the value for a property follows:
+ * 1. matched path level configuration;
+ * 2. cluster level configuration.
+ */
+public final class PathConfiguration implements AlluxioConfiguration {
+  private final AlluxioConfiguration mClusterConf;
+  private final Map<String, AlluxioConfiguration> mPathConf;
+  private final AlluxioURI mPath;
+  private final Filter mFilter;
+
+  /**
+   * Filter out the path level configuration for a specific key.
+   */
+  private interface Filter {
+    AlluxioConfiguration filter(PropertyKey key);
+  }
+
+  private final class PathFilter implements Filter {
+    @Override
+    public AlluxioConfiguration filter(PropertyKey key) {
+      AlluxioConfiguration conf = mPathConf.get(mPath.getPath());
```

Comment:
this lookup doesn't look right. path defaults should apply to entire sub-tree, not just the path for which it was set. i think we need to re-work the data structure for look up

Reply:
I see. For this PR, let's focus on the overall interface and class design, the details of implementation can be updated.